### ISSUE TITLE
Node process does not abort when it is out of memory

### ISSUE THREAD

**Issue by doganyazar** [RisingStack/docker-node#1]
This is a copy of an issue I have created for alphine-node here first: https://github.com/mhart/alpine-node/issues/44
I have experienced the same problem using your images too so just wanted to see if you have any idea.

My node process does not exit (aborted) when it is out of memory. Instead it just prints

```
<--- Last few GCs --->

    1095 ms: Mark-sweep 27.4 (62.7) -> 21.2 (59.4) MB, 33.0 / 0 ms (+ 5.5 ms in 42 steps since start of marking, biggest step 0.4 ms) [allocation failure] [GC in old space requested].
    1114 ms: Mark-sweep 21.2 (59.4) -> 21.2 (59.4) MB, 19.1 / 0 ms [allocation failure] [GC in old space requested].
    1132 ms: Mark-sweep 21.2 (59.4) -> 20.9 (59.4) MB, 17.6 / 0 ms [last resort gc].
    1149 ms: Mark-sweep 20.9 (59.4) -> 20.9 (59.4) MB, 17.5 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1767b25c9e59 <JS Object>
    2: _repeat [/usr/src/app/memory.js:~10] [pc=0x54102d5f9b1] (this=0x3cc32928c999 <a Timeout with map 0x34877b623bb1>)
    3: _onTimeout(aka wrapper) [timers.js:417] [pc=0x54102d5d7b3] (this=0x3cc32928c999 <a Timeout with map 0x34877b623bb1>)
    4: tryOnTimeout(aka tryOnTimeout) [timers.js:224] [pc=0x54102d5d4e3] (this=0x1767b2504189 <undefined>,timer=0x3cc32928c999 <a Timeout with map 0x348...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
and then starts using 100% CPU. 
This is a serious safety issue. Is this expected due to difference with glibc and musl?
```

When I try it with node:argon container exits as expected after printing the same information as above.

To mimic the behaviour, I have created a Dockerfile and a js file that leaks memory:
[memory.zip](https://github.com/RisingStack/docker-node/files/314399/memory.zip)

**Comment by gergelyke** [comment]
Hello @doganyazar !

thanks for reporting, checking now!

**Comment by doganyazar** [comment]
Hi @gergelyke, any follow up on this issue? 

We did some more digging and it seems to be caused by several issues together. I tried to list them below:
- musl handles abort differently compared to glibc. Created issue on their mail list: http://www.openwall.com/lists/musl/2016/06/18/4 (click thread-next to see their answers)
- Docker running processes as PID 1 (and hence process behaves differently)
- Node does not seem to trigger explicit SIGABRT handler. Created issue: https://github.com/nodejs/node/issues/7341

**Comment by karlbohlmark** [comment]
Just FYI: I looked furthered into this and noticed that it was caused by musl libc abort() not actually causing termination of PID 1 processes. I talked to the guys on that project and Rich made a fix (https://git.musl-libc.org/cgit/musl/commit/?id=0c8bc102f287d3993751d80ba2dffb01e0c8bc7f) which is released in musl version 1.1.15.

I guess this will resolve itself when the next Alpine release comes out.
